### PR TITLE
VultureBear: Make use of the vulture API

### PR DESCRIPTION
**Pull request by RJ722**
Refactor VultureBear to make use of the vulture API rather than using it as a 
command line tool.

Closes: https://github.com/coala/coala-bears/issues/1805

### Checklist

- [X] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [X] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: bears/python/VultureBear.py [hunk 1]
```diff
@@ -1,25 +1,56 @@
...
+
+class Unused(Item):
+
+    def __new__(cls, item, message):
+        item.message = message
+        return item
+
+
+def retrieve_unused(filenames):
+    import vulture
```
Comment (RJ722) [inline bears/python/VultureBear.py]: If the `import vulture` statement is moved at the top of the file, an error occurs:
```
Traceback (most recent call last):
  File "/home/rahul/.virtual_envs/coala_dev/lib/python3.4/site-packages/coalib/bears/Bear.py", line 282, in execute
    return [] if result is None else list(result)
  File "/home/rahul/Desktop/code/coala-bears/bears/python/VultureBear.py", line 68, in run
    unused = retrieve_unused(filenames)
  File "/home/rahul/Desktop/code/coala-bears/bears/python/VultureBear.py", line 34, in retrieve_unused
    vulture = vulture.Vulture()
UnboundLocalError: local variable 'vulture' referenced before assignment
```
:confused:

File: bears/python/VultureBear.py [hunk 2]
```diff
@@ -32,14 +63,10 @@ def run(self):
...
-        stdout_output, _ = run_shell_command(
-            (self.EXECUTABLE,) +
-            tuple(filename for filename in self.file_dict.keys()),
-            cwd=self.get_config_dir())
-
-        for match in re.finditer(self.OUTPUT_REGEX, stdout_output):
-            groups = match.groupdict()
+        filenames = list(self.file_dict.keys())
+        unused = retrieve_unused(filenames)
+        for unused_entity in unused:
```
Comment (RJ722) [inline bears/python/VultureBear.py]: TODO: Since unused imports are detected with 100% accuracy, implement a patch yielding result which automatically removes the statement.

File: bears/python/VultureBear.py (hunk 1 above)
Comment (jayvdb) [inline bears/python/VultureBear.py]: That is because you re-assign a different value to `vulture` later:

> vulture = vulture.Vulture()

Bam!  python decides you want a local variable, and not use the global variable.

File: bears/python/VultureBear.py (hunk 2 above)
Comment (jayvdb) [inline bears/python/VultureBear.py]: Do you mean add code for removing unused imports into vulture?

If so, we have [autoflake](https://github.com/myint/autoflake/) for that, and a coala GSOC project will be doing crazy enhancements like https://gitlab.com/coala/GSoC-2017/issues/216

File: bears/python/VultureBear.py [hunk 3]
```diff
@@ -1,25 +1,45 @@
...
 
+class Unused(Item):
+
+    def __new__(cls, item, message):
+        item.message = message
+        return item
+
+
+def retrieve_unused(filenames):
+
```
Comment (jayvdb) [inline bears/python/VultureBear.py]: Needs a docstring, and possible this function should be private by prefixing name with `_`

File: bears/python/VultureBear.py [hunk 4]
```diff
@@ -1,25 +1,45 @@
...
+    for filename in filenames:
+        vulture = Vulture()
+        vulture.scavenge([filename])
+        for item in sorted(
+                vulture.unused_funcs + vulture.unused_imports +
+                vulture.unused_props + vulture.unused_vars +
+                vulture.unused_attrs, key=file_lineno):
+            message = 'Unused %s: %s' % (item.typ, item)
+            unused.append(Unused(item, message))
```
Comment (jayvdb) [inline bears/python/VultureBear.py]: Why not build `Result` objects here, instead of creating and build `Item` objects?

File: bears/python/VultureBear.py [hunk 5]
```diff
@@ -1,25 +1,45 @@
...
+
+def retrieve_unused(filenames):
+
+    def file_lineno(item):
+        return (item.filename.lower(), item.lineno)
+    unused = []
+
+    for filename in filenames:
+        vulture = Vulture()
+        vulture.scavenge([filename])
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: This scans unused code file by file. What the bear previously did can be achieved by

```
vulture = Vulture()
vulture.scavenge(filenames)
for item in sorted(...):
    ...
```

File: bears/python/VultureBear.py (hunk 4 above)
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: I agree, there's no need for the `Unused` class.

File: bears/python/VultureBear.py [hunk 6]
```diff
@@ -1,25 +1,45 @@
-import re
 from shutil import which
-
+from vulture import Item, Vulture
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: I'd sort the imports alphabetically, i.e. coalib, dependency_management, vulture

File: bears/python/VultureBear.py [hunk 7]
```diff
@@ -1,25 +1,45 @@
...
+    unused = []
+
+    for filename in filenames:
+        vulture = Vulture()
+        vulture.scavenge([filename])
+        for item in sorted(
+                vulture.unused_funcs + vulture.unused_imports +
+                vulture.unused_props + vulture.unused_vars +
+                vulture.unused_attrs, key=file_lineno):
+            message = 'Unused %s: %s' % (item.typ, item)
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: I recommend using `str.format` instead of `%`.

File: bears/python/VultureBear.py [hunk 8]
```diff
@@ -1,25 +1,45 @@
...
     LICENSE = 'AGPL-3.0'
     ASCIINEMA_URL = 'https://asciinema.org/a/82256'
     CAN_DETECT = {'Unused Code'}
 
-    EXECUTABLE = 'vulture'
-    OUTPUT_REGEX = re.compile(
-        r'(?P<filename>.*):(?P<line>.*):\s*(?P<message>.*)')
-
     @classmethod
     def check_prerequisites(cls):
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: The check needs to be updated. Instead of `which`, we need to test importing vulture:

above:
```
try:
    from vulture import Vulture
except ImportError:
    Vulture = None
```

below:

```
@classmethod
def check_prerequisites(cls):
    return ('...' if Vulture is None else True)
```

Or is listing vulture in the `REQUIREMENTS` already enough @jayvdb ?

File: bears/python/VultureBear.py [hunk 9]
```diff
@@ -32,14 +52,10 @@ def run(self):
 
         See <https://bitbucket.org/jendrikseipp/vulture> for more information.
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: please update repo link

File: bears/python/VultureBear.py (hunk 8 above)
Comment (RJ722) [inline bears/python/VultureBear.py]: I uninstalled vulture and then ran: 
```
coala -b VultureBear -f *.py -L DEBUG --flush-cache
```
It gave a misleading error message: 
```
[WARNING][16:58:24] No bears matching 'VultureBear' were found. Make sure you have coala-bears installed or you have typed the name correctly.
```

After re-installing vulture, the same command worked. :confused:

Comment (RJ722) [inline bears/python/VultureBear.py]: I think that whenever coala finds a Bear whose requirements aren't installed yet, it  _deactivates_ that Bear.

If this is the case, then why do we return a message from `check_prerequisites` about how to get dependencies installed. 😕

File: bears/python/VultureBear.py [hunk 10]
```diff
@@ -1,45 +1,52 @@
...
-
-        for match in re.finditer(self.OUTPUT_REGEX, stdout_output):
-            groups = match.groupdict()
-            yield Result.from_values(origin=self,
-                                     message=groups['message'],
-                                     file=groups['filename'],
-                                     line=int(groups['line']))
+        filenames = list(self.file_dict.keys())
+        unused = _retrieve_unused(filenames)
+        for unused_entity in unused:
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: no need for `unused` variable. The code `for unused_entity in _retrieve_unused(filenames):` is both shorter and signals that the result of the function call is not used anywhere else.

File: bears/python/VultureBear.py [hunk 11]
```diff
@@ -1,45 +1,52 @@
...
-
 from coalib.bears.GlobalBear import GlobalBear
-from dependency_management.requirements.PipRequirement import PipRequirement
-from coalib.misc.Shell import run_shell_command
 from coalib.results.Result import Result
+from dependency_management.requirements.PipRequirement import PipRequirement
+from vulture import Vulture
+
+
+def _retrieve_unused(filenames):
```
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: `_retrieve_unused` is missing an object and "retrieve" suggests that the analysis has already been done. I'd call the function `_find_unused_code` instead.

File: bears/python/VultureBear.py (hunk 10 above)
Comment (jendrikseipp) [inline bears/python/VultureBear.py]: We could even just do the following, I guess:

```
return _find_unused_code(filenames)
```

**Comment by jendrikseipp** [comment]
Unfortunately, I don't get notified when you push commits. Can you please ping me when you're done with the changes?

File: bears/python/VultureBear.py [hunk 12]
```diff
@@ -1,45 +1,45 @@
...
-        return ('Vulture is missing. Make sure to install it using '
-                '`pip3 install vulture`.'
-                if which('vulture') is None else True)
-
     def run(self):
         """
         Check Python code for unused variables and functions using `vulture`.
 
-        See <https://bitbucket.org/jendrikseipp/vulture> for more information.
+        See <https://github.com/jendrikseipp/vulture> for more information.
```
Comment (jayvdb) [inline bears/python/VultureBear.py]: This is an unrelated change, and should be a separate commit, or even a separate PR, and as you will be modifying this line anyway, you can move it to a `SEE_MORE` attribute (c.f. https://github.com/coala/coala-bears/issues/1136)

Comment (RJ722) [inline bears/python/VultureBear.py]: Raised https://github.com/coala/coala-bears/pull/1835

**Comment by meetmangukiya** [comment]
ack 7f3fe6fda39332a380762c36ba800516f14a0694

**Comment by meetmangukiya** [comment]
@rultor merge
